src/olympia/api/tests/test_authentication.py

```diff
@@ -214,32 +214,90 @@ def test_success(self):
+                WebTokenAuthentication.auth_header_prefix))
+        with self.assertRaises(AuthenticationFailed):
+            self.auth.authenticate(request)
+
+    def test_no_token(self):
+        request = self.factory.post('/api/v3/whatever/')
+        self.auth.authenticate(request) is None
+
+    def test_expired_token(self):
+        old_date = datetime.now() - timedelta(
+            seconds=settings.SESSION_COOKIE_AGE + 1)
+        with freeze_time(old_date):
+            token = self.client.generate_api_token(self.user)
+        with self.assertRaises(AuthenticationFailed):
+            self._authenticate(token)
+
+    def test_still_valid_token(self):
+        not_so_old_date = datetime.now() - timedelta(
+            seconds=settings.SESSION_COOKIE_AGE - 30)
+        with freeze_time(not_so_old_date):
+            token = self.client.generate_api_token(self.user)
+        assert self._authenticate(token)[0] == self.user
+
+    def test_bad_token(self):
+        token = 'garbage'
+        with self.assertRaises(AuthenticationFailed):
+            self._authenticate(token)
 
-    def test_no_user_id(self):
+    def test_user_id_is_none(self):
         token = self.client.generate_api_token(self.user, user_id=None)
         with self.assertRaises(AuthenticationFailed):
             self._authenticate(token)
 
+    def test_no_user_id_in_payload(self):
+        data = {
+            'secret': self.user.get_session_auth_hash(),
+        }
+        token = signing.dumps(data, salt=WebTokenAuthentication.salt)
+        with self.assertRaises(AuthenticationFailed):
+            self._authenticate(token)
+
+    def test_no_secret_in_payload(self):
+        data = {
+            'user_id': self.user.pk,
+        }
+        token = signing.dumps(data, salt=WebTokenAuthentication.salt)
+        with self.assertRaises(AuthenticationFailed):
+            self._authenticate(token)
+
     def test_user_deleted(self):
         self.user.anonymize()
         token = self.client.generate_api_token(self.user)
         with self.assertRaises(AuthenticationFailed):
             self._authenticate(token)
 
-    def test_username_changes(self):
-        token = self.client.generate_api_token(self.user)
-        self.user.update(username='different')
-        user, _ = self._authenticate(token)
-        assert user == self.user
-
     def test_invalid_user_not_found(self):
         token = self.client.generate_api_token(self.user, user_id=-1)
         with self.assertRaises(AuthenticationFailed):
             self._authenticate(token)
+
+    def test_invalid_user_other_user(self):
+        user2 = user_factory(read_dev_agreement=datetime.now())
+        token = self.client.generate_api_token(self.user, user_id=user2.pk)
+        with self.assertRaises(AuthenticationFailed):
+            self._authenticate(token)
+
+    def test_wrong_auth_id(self):
+        token = self.client.generate_api_token(self.user)
```

Comment:
I think it might help to add a test similar to this that makes sure the `user_id` is included in the token since frontend will be relying on it.